datumaro/components/extractor_tfds.py

```diff
@@ -0,0 +1,280 @@
+    Any, Callable, Dict, Iterator, Optional, Sequence, Set, Tuple, Union,
+)
+import itertools
+import logging as log
+import os.path as osp
+
+from attr import attrib, attrs
+
+from datumaro.components.annotation import (
+    AnnotationType, Bbox, Label, LabelCategories,
+)
+from datumaro.components.extractor import (
+    CategoriesInfo, DatasetItem, IExtractor,
+)
+from datumaro.components.media import ByteImage
+from datumaro.util.tf_util import import_tf
+
+try:
+    tf = import_tf()
+    import tensorflow_datasets as tfds
+except ImportError:
+    log.debug("Unable to import TensorFlow or TensorFlow Datasets. " \
+        "Dataset downloading via TFDS is disabled.")
+    TFDS_EXTRACTOR_AVAILABLE = False
+else:
+    TFDS_EXTRACTOR_AVAILABLE = True
+
+@attrs(auto_attribs=True)
+class _TfdsAdapter:
+    metadata_transformers: Sequence[
+        Callable[['tfds.core.DatasetBuilder', CategoriesInfo], None]]
+    data_transformers: Sequence[Callable[[Any, DatasetItem], None]]
+    id_generator: Callable[[Any], str] = attrib(default=None, kw_only=True)
+
+    def transform_metadata(self,
+        tfds_builder: 'tfds.core.DatasetBuilder', categories: CategoriesInfo,
+    ) -> None:
+        for t in self.metadata_transformers:
+            t(tfds_builder, categories)
+
+    def transform_data(self, tfds_example: Any, item: DatasetItem) -> None:
+        for t in self.data_transformers:
+            t(tfds_example, item)
+
+@attrs(auto_attribs=True)
+class _SetLabelCategoriesFromClassLabelFeature:
+    feature_path: Union[str, Tuple[str, ...]]
+
+    def __call__(self,
+        tfds_builder: 'tfds.core.DatasetBuilder', categories: CategoriesInfo,
+    ) -> None:
+        assert AnnotationType.label not in categories
+        if isinstance(self.feature_path, str):
+            feature_connector = tfds_builder.info.features[self.feature_path]
+        else:
+            feature_connector = tfds_builder.info.features
+
+            for segment in self.feature_path:
+                assert isinstance(feature_connector, (
+                    tfds.features.FeaturesDict, tfds.features.Sequence,
+                ))
+
+                if isinstance(feature_connector, tfds.features.Sequence):
+                    assert segment == 'feature'
+                    feature_connector = feature_connector.feature
+                else:
+                    feature_connector = feature_connector[segment]
+
+        assert isinstance(feature_connector, tfds.features.ClassLabel)
+        categories[AnnotationType.label] = LabelCategories.from_iterable(
+            feature_connector.names)
+
+@attrs(auto_attribs=True)
+class _SetImageFromImageFeature:
+    feature_name: str
+    filename_feature_name: Optional[str] = attrib(default=None)
+
+    def __call__(self, tfds_example: Any, item: DatasetItem) -> None:
+        if self.filename_feature_name:
+            filename = tfds_example[self.filename_feature_name].numpy(
+            ).decode('UTF-8')
```

Comment:
```suggestion
            filename = tfds_example[self.filename_feature_name].numpy() \
                .decode('UTF-8')
```
The formatting is strange here.

Reply:
Yeah, I don't know what I was thinking. Fixed.